README.md

```diff
@@ -1,488 +1,742 @@
+Metas are arbitrary key/value pairs that can be associated with any node. Metas can be used to store metadata about a
+node, such as the location in the source text of the grammar element it represents, or the data type of the value
+returned by an expression.
 
-1. the constructors of record types must validate the minimum arity of variadic elements at runtime
-1. the minimum arity of variadic elements is always zero (and PIG fails the build otherwise). 
+## Traversing & Transforming Trees
 
-An example of a domain-level record:
+PIG generates four different types of tree traversals for performing various tasks. Each is described in the
+sub-headings of this section.
 
-```
-(define demo_domain 
-    (domain
-        ...
-        (record mathematician
-            (first_name symbol)
-            (last_name symbol)
-            (age int)
-            (games_of_life_conceived int)
-        ...)))
-
-// Example s-expression representation:
-(programmer
-    (first_name John)
-    (last_name Conway)
-    (age 82)
-    (games_of_life_conceived 1))
-```
+For the examples below we will use an `enhanced_calculator_ast`. This is like the `calculator_ast` from before but also
+includes a `let` and `variable` expressions.
 
-An example of a sum variant record:    
+```text
+(define enhanced_calculator_ast
+        (domain
+            (sum expr
+                 (lit value::int)
+                 (binary op::operator left::expr right::expr)
+                 (let name::symbol value::expr body::expr)
+                 (variable name::symbol) 
+             )
 
-```
-(define partiql 
-    (domain
-        ...
-        (sum expr
-            ...
-            (select
-                (project projection)
-                (from from_source)
-                (where (? expr))
-                (group_by (? group_specificaiton))
-                (having (? expr))
-                (limit (? expr)))
-            ...)
-        ...)))
+            (sum operator (plus) (minus) (times) (divide) (modulo))
+        )
+)
 ```
 
-##### Naming Conventions
+### `VisitorTransform`
 
-All nameable components (type domains, types, elements, etc) of a type universe file should be named using the  
-[`snake_case`](https://en.wikipedia.org/wiki/Snake_case) naming convention.
+The PIG-generated `VisitorTransform` base classes are by far the most commonly used type of tree traversal. Using a
+`VisitorTransform`, it is possible to implement many kinds of compiler optimizations, static type inference, expression
+normalization, and syntactic de-sugaring.
 
-The language target may convert this to another naming convention.  For instance, the Kotlin target converts
-`snake_case` to [`camelCase` or `PascalCase`](https://en.wikipedia.org/wiki/Camel_case) as appropriate to 
-the context in order to provide an idiomatic API to the generated code.
+By default, every type domain gets one `VisitorTransform` capable of
```

Comment:
Incomplete sentence. "...capable of..."

Reply:
cleaned up several other sentences around here as well.